From 6d22c9abc29d43e919e819ff004fcd84a90de60b Mon Sep 17 00:00:00 2001
From: Nero <nero@w1r3.net>
Date: Sat, 23 Sep 2023 11:50:04 +0000
Subject: [PATCH] install: Fix chown resetting suid/sgid bits from chmod
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Since Linux 2.2.13, chown(2) resets the suid/gid bits for all users.
This patch changes the ordering so that chmod gets called after chown.

This behavior follows GNU coreutils.

Signed-off-by: Nero <nero@w1r3.net>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
 coreutils/install.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/coreutils/install.c b/coreutils/install.c
index c0f1c538a..00f8be87e 100644
--- a/coreutils/install.c
+++ b/coreutils/install.c
@@ -244,6 +244,15 @@ int install_main(int argc, char **argv)
 			}
 		}
 
+		/* Set the user and group id */
+		/* (must be before chmod, or else chown may clear suid/gid bits) */
+		if ((opts & (OPT_OWNER|OPT_GROUP))
+		 && lchown(dest, uid, gid) == -1
+		) {
+			bb_perror_msg("can't change %s of %s", "ownership", dest);
+			ret = EXIT_FAILURE;
+		}
+
 		/* Set the file mode (always, not only with -m).
 		 * GNU coreutils 6.10 is not affected by umask. */
 		if (chmod(dest, mode) == -1) {
@@ -254,13 +263,6 @@ int install_main(int argc, char **argv)
 		if (use_default_selinux_context)
 			setdefaultfilecon(dest);
 #endif
-		/* Set the user and group id */
-		if ((opts & (OPT_OWNER|OPT_GROUP))
-		 && lchown(dest, uid, gid) == -1
-		) {
-			bb_perror_msg("can't change %s of %s", "ownership", dest);
-			ret = EXIT_FAILURE;
-		}
  next:
 		if (ENABLE_FEATURE_CLEAN_UP && isdir)
 			free(dest);
